src/prefab_manager.zig

```diff
@@ -0,0 +1,197 @@
+const rt = @import("renderer/renderer_types.zig");
+const IdLocal = @import("variant.zig").IdLocal;
+
+const assert = std.debug.assert;
+const d3d12 = zwin32.d3d12;
+const zcgltf = zmesh.io.zcgltf;
+
+const PrefabHashMap = std.AutoHashMap(IdLocal, flecs.Entity);
+
+pub const PrefabManager = struct {
+    prefab_hash_map: PrefabHashMap,
+    is_a: flecs.Entity,
+
+    pub fn init(world: *flecs.World, allocator: std.mem.Allocator) PrefabManager {
+        return PrefabManager{
+            .prefab_hash_map = PrefabHashMap.init(allocator),
+            .is_a = flecs.Entity.init(world.world, flecs.c.Constants.EcsIsA),
+        };
+    }
+
+    pub fn deinit(self: *@This()) void {
+        self.prefab_hash_map.deinit();
+    }
+
+    pub fn loadPrefabFromGLTF(
+        self: *@This(),
+        path: [:0]const u8,
+        world: *flecs.World,
+        gfxstate: *gfx.D3D12State,
+        allocator: std.mem.Allocator
+    ) !flecs.Entity {
+        const path_id = IdLocal.init(path);
+        var existing_prefab = self.prefab_hash_map.get(path_id);
+        if (existing_prefab) |prefab| {
+            return prefab;
+        }
+
+        const data = try zmesh.io.parseAndLoadFile(path);
+        defer zmesh.io.freeData(data);
+
+        assert(data.scenes_count == 1);
+        const scene = &data.scenes.?[0];
+        assert(scene.nodes_count == 1);
+
+        var arena_state = std.heap.ArenaAllocator.init(allocator);
+        defer arena_state.deinit();
+        const arena = arena_state.allocator();
+
+        var prefab = world.newPrefab(path);
+        prefab.setOverride(fd.Position.init(0, 0, 0));
+        prefab.setOverride(fd.EulerRotation{});
+        prefab.setOverride(fd.Scale.createScalar(1));
+        prefab.setOverride(fd.Transform{});
+        prefab.setOverride(fd.Forward{});
+        prefab.setOverride(fd.Dynamic{});
+        self.parseNode(scene.nodes.?[0], prefab, world, gfxstate, arena);
+
+        self.prefab_hash_map.put(path_id, prefab) catch unreachable;
+
+        return prefab;
+    }
+
+    pub fn getPrefabByPath(self: *@This(), path: [:0]const u8) ?flecs.Entity {
+        const path_id = IdLocal.init(path);
+        var existing_prefab = self.prefab_hash_map.get(path_id);
+        if (existing_prefab) |prefab| {
+            return prefab;
+        }
+
+        return null;
+    }
+
+    pub fn instantiatePrefab(self: @This(), world: *flecs.World, prefab: flecs.Entity) flecs.Entity {
+        const entity = world.newEntity();
+        entity.addPair(self.is_a, prefab);
+        return entity;
+    }
+
+    fn parseNode(
+        self: *@This(),
+        node: *zcgltf.Node,
+        parent_entity: flecs.Entity,
+        world: *flecs.World,
+        gfxstate: *gfx.D3D12State,
+        arena: std.mem.Allocator,
+    ) void {
+        // Set parent
+        var entity = world.newPrefab(node.name);
+        entity.addPair(flecs.c.Constants.EcsChildOf, parent_entity);
+        entity.setOverride(fd.Forward{});
```

Comment:
Same here, there could be "logic" nodes that aren't really world-spacey. But also for the future :)